Add Jenkins build pipeline

djbingham opened the pull request:
https://eaflood.atlassian.net/browse/PSD-81

Commit:
Create Jenkins build pipeline

https://eaflood.atlassian.net/browse/PSD-81

Commit:
Initial Jenkinsfile

Copied from ffc-demo-web with image and repo names changed.

Commit:
Correct docker-compose commands in Jenkinsfile

Commit:
Correct test container image name

Commit:
Correct production image name and build in Jenkins

Commit:
Add jenkins-aws.yaml to Helm chart

Commit:
Correct path to Jenkins Helm install values file

Commit:
Add api-gateway to name of PR deployments

Commit:
Set chart name correctly for AWS Jenkins deploy

Commit:
Remove ingress settings from Jenkins PR deploy

Commit:
Remove service from Jenkins-deployed Helm values

Commit:
Switch `up` to `run` in Jenkins test stage

Seems a more natural command to use for running a test, as opposed to
starting a long-lived service.

johnwatson484 commented on Jenkinsfile:
```diff
@@ -0,0 +1,131 @@
+
+def getMergedPrNo() {
+    def mergedPrNo = sh(returnStdout: true, script: "git log --pretty=oneline --abbrev-commit -1 | sed -n 's/.*(#\\([0-9]\\+\\)).*/\\1/p'").trim()
+    return mergedPrNo ?: ''
```
should be `return mergedPrNo ? "pr$mergedPrNo" : ''` as the PR number returned is only the number, but the actual namespace that needs to be undeployed later has the "pr" prefix.

djbingham replied:
Ah yes, I can see how that causes an issue. I don't like returning a non-numeric string from a function whose name indicates that a number should be returned (or indeed, setting a variable named `mergedPrNo` to a non-numeric value).

To me, it would feel more natural to add the `pr` prefix as required (in the undeploy stage):
```
undeployPR(kubeCredsId, imageName, "pr$mergedPrNo")
```

Would that be ok, or do you think it's better to just match the pipelines that have already been done?

I've pushed that change so you can see it in context. Happy to revert if you'd rather just match the other pipelines here than go back and change others to match this.

djbingham replied:
Looking at the number of pipelines already in review, I think it makes more sense to stick with the solution already in use for now. We'll be abstracting the Jenkinsfile functions to a library next sprint anyway, so that's a good opportunity to make improvements.

I've pushed the change you suggested.

johnwatson484 commented on Jenkinsfile:
```diff
@@ -0,0 +1,131 @@
+}
+
+def buildTestImage(name, suffix) {
+  sh 'docker image prune -f'
```
I don't think this is necessary now we've added it as a nightly job.  Running this in a pipeline risks breaking any other pipelines that may be running.

djbingham replied:
Thanks, I've removed that.

johnwatson484 commented on Jenkinsfile:
```diff
@@ -0,0 +1,131 @@
+    sh "docker-compose -p $name-$suffix -f docker-compose.test.yaml down -v"
+    junit 'test-output/junit.xml'
+    // clean up files created by node/ubuntu user that cannot be deleted by jenkins. Note: uses global environment variable
+    sh "docker run -u node --mount type=bind,source=$WORKSPACE/test-output,target=/usr/src/app/test-output $name rm -rf test-output/*"
```
the workspace directory needs to be in quotes.

eg source='$WORKSPACE/test-output'

djbingham replied:
Quotes added, thanks.

johnwatson484 commented on Jenkinsfile:
```diff
@@ -0,0 +1,131 @@
+  }
+  if (pr != '') {
+    stage('Helm install') {
+      def extraCommands = "--values ./helm/$repoName/jenkins-aws.yaml --set name=ffc-demo-api-gateway-$containerTag"
```
There's no need to set the name Helm value here, it's set for the web as the ingress needs it but it can just stay as the default in the values file here.

djbingham replied:
Good point, I've removed that.

johnwatson484 reviewed (changes requested):
There's a couple of changes needed to fix some of the issues that were found in the last iteration of the pipeline.  I've added comments to detail what needs to be changed.

Commit:
Stop pruning Docker images during Jenkins build

Images are cleaned up nightly. Pruning during the pipeline risks
breaking other pipelines running in parallel.

Commit:
Quote directory path in Jenkinsfile to prevent errors

Commit:
Remove unnecessary Helm chart name override

Commit:
Correct PR tag used to undeploy from Jenkins

Commit:
Set PR tag in Jenkinsfile the same as other repos

johnwatson484 reviewed (approved):
LGTM
